Jenkinsfile

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env groovy
+def results = [:]
+
+pipeline {
+  agent none
+  environment {
+    BASE_DIR="src/github.com/elastic/apm-agent-python"
+  }
+  options {
+    timeout(time: 1, unit: 'HOURS') 
+    buildDiscarder(logRotator(numToKeepStr: '20', artifactNumToKeepStr: '20', daysToKeepStr: '30'))
+    timestamps()
+    ansiColor('xterm')
+    disableResume()
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+  }
+  parameters {
+    booleanParam(name: 'Run_As_Master_Branch', defaultValue: false, description: 'Allow to run any steps on a PR, some steps normally only run on master branch.')
+    booleanParam(name: 'doc_ci', defaultValue: true, description: 'Enable build docs.')
+  }
+  stages {
+    stage('Initializing'){
+      agent { label 'docker && linux && immutable' }
+      options { skipDefaultCheckout() }
+      environment {
+        HOME = "${env.WORKSPACE}"
+        PATH = "${env.PATH}:${env.WORKSPACE}/bin"
+        ELASTIC_DOCS = "${env.WORKSPACE}/elastic/docs"
+      }
+      stages {
+        /**
+         Checkout the code and stash it, to use it on other stages.
+        */
+        stage('Checkout') {
+          steps {
+            deleteDir()
+            gitCheckout(basedir: "${BASE_DIR}")
+            stash allowEmpty: true, name: 'source', useDefaultExcludes: false
+          }
+        }
+        /**
+         Build the project from code..
+        */
+        stage('Build') {
+          steps {
+            deleteDir()
+            unstash 'source'
+            dir("${BASE_DIR}"){
+              sh """
+              ./tests/scripts/docker/cleanup.sh
+              ./tests/scripts/docker/isort.sh
+              """
+              sh """
+              ./tests/scripts/docker/cleanup.sh
+              ./tests/scripts/docker/black.sh
+              """
+            }
+          }
+        }
+        /**
+         Execute unit tests.
+        */
+        stage('Test') {
+          steps {
+            deleteDir()
+            unstash 'source'
+            script {
+              def parallelStages = [:]
+              getPythonVersions().findAll{ it.startsWith('pypy') }.each{ py ->
```

Comment:
@kuisathaverat Are the stages run in parallel on the same worker or different jenkins workers?

Reply:
I am experimenting with some solutions to make it efficient and UI friendly

* Every test on a stage running on a node in parallel https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-python%2Fapm-agent-python-mbp/detail/PR-360/10/pipeline
* this one group python version tests on a single node and run then in parallel https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-python%2Fapm-agent-python-mbp/detail/PR-360/15/pipeline is much better in UI but it is slow

Reply:
I personally would always take the speed over nice UI here :-)

Reply:
There might be some issues with running tests in parallel on the same node, especially tests that interact with an external resource (like the Elasticsearch or MongoDB tests). Let me know if you hit any weird test failures and I'll investigate.

Reply:
I meant, I spin one node per python version in parallel and execute python test group sequentially on the same node, with a cleanup before each one. This has the advantage that only builds ones the Python Docker image for a node, but it is still a little slow I checking how to improve it.

Reply:
also, the UI is friendly (mostly)